pkg/ingester/user_state.go

```diff
@@ -114,59 +115,104 @@ func newLabelSetCounter(limiter *Limiter) *labelSetCounter {
 }
 
 func (m *labelSetCounter) canAddSeriesForLabelSet(ctx context.Context, u *userTSDB, metric labels.Labels) error {
-	return m.limiter.AssertMaxSeriesPerLabelSet(u.userID, metric, func(set validation.LimitsPerLabelSet) (int, error) {
-		s := m.shards[util.HashFP(model.Fingerprint(set.Hash))%numMetricCounterShards]
+	return m.limiter.AssertMaxSeriesPerLabelSet(u.userID, metric, func(allLimits []validation.LimitsPerLabelSet, limit validation.LimitsPerLabelSet) (int, error) {
+		s := m.shards[util.HashFP(model.Fingerprint(limit.Hash))%numMetricCounterShards]
 		s.RLock()
-		if r, ok := s.valuesCounter[set.Hash]; ok {
+		if r, ok := s.valuesCounter[limit.Hash]; ok {
 			defer s.RUnlock()
 			return r.count, nil
 		}
 		s.RUnlock()
 
 		// We still dont keep track of this label value so we need to backfill
-		return m.backFillLimit(ctx, u, set, s)
+		return m.backFillLimit(ctx, u, false, allLimits, limit, s)
 	})
 }
 
-func (m *labelSetCounter) backFillLimit(ctx context.Context, u *userTSDB, limit validation.LimitsPerLabelSet, s *labelSetCounterShard) (int, error) {
+func (m *labelSetCounter) backFillLimit(ctx context.Context, u *userTSDB, forceBackfill bool, allLimits []validation.LimitsPerLabelSet, limit validation.LimitsPerLabelSet, s *labelSetCounterShard) (int, error) {
+	s.Lock()
+	// If not force backfill, use existing counter value.
+	if !forceBackfill {
+		if r, ok := s.valuesCounter[limit.Hash]; ok {
+			s.Unlock()
+			return r.count, nil
+		}
+	}
+
+	defer s.Unlock()
 	ir, err := u.db.Head().Index()
 	if err != nil {
 		return 0, err
 	}
 
 	defer ir.Close()
 
-	s.Lock()
-	defer s.Unlock()
-	if r, ok := s.valuesCounter[limit.Hash]; !ok {
-		postings := make([]index.Postings, 0, len(limit.LabelSet))
-		for _, lbl := range limit.LabelSet {
-			p, err := ir.Postings(ctx, lbl.Name, lbl.Value)
-			if err != nil {
-				return 0, err
-			}
-			postings = append(postings, p)
-		}
+	numSeries := u.db.Head().NumSeries()
+	totalCount, err := getCardinalityForLimitsPerLabelSet(ctx, numSeries, ir, allLimits, limit)
+	if err != nil {
+		return 0, err
+	}
 
-		p := index.Intersect(postings...)
+	s.valuesCounter[limit.Hash] = &labelSetCounterEntry{
+		count:  totalCount,
+		labels: limit.LabelSet,
+	}
+	return totalCount, nil
+}
 
-		totalCount := 0
-		for p.Next() {
-			totalCount++
+func getCardinalityForLimitsPerLabelSet(ctx context.Context, numSeries uint64, ir tsdb.IndexReader, allLimits []validation.LimitsPerLabelSet, limit validation.LimitsPerLabelSet) (int, error) {
+	// Easy path with explicit labels.
+	if limit.LabelSet.Len() > 0 {
```

Comment:
Should we document in the label set config that a empty labelset is the "default" and will match all other series?

Reply:
Updated in the latest commit